internal/jobcontainers/process.go

```diff
@@ -0,0 +1,171 @@
+	return nil
+}
+
+// Stdio returns the stdio pipes of the process
+func (p *JobProcess) Stdio() (io.Writer, io.Reader, io.Reader) {
+	return p.stdin, p.stdout, p.stderr
+}
+
+// Signal sends a signal to the process.
+func (p *JobProcess) Signal(ctx context.Context, options interface{}) (bool, error) {
+	if p.exited() {
+		return false, errors.New("signal not sent. process already exited")
+	}
+
+	p.procLock.Lock()
+	defer p.procLock.Unlock()
+	if err := windows.GenerateConsoleCtrlEvent(windows.CTRL_BREAK_EVENT, uint32(p.Pid())); err != nil {
+		return false, fmt.Errorf("failed to send signal: %s", err)
+	}
+	return true, nil
+}
+
+// CloseStdin closes the stdin pipe of the process.
+func (p *JobProcess) CloseStdin(ctx context.Context) error {
+	p.stdioLock.Lock()
+	defer p.stdioLock.Unlock()
+	return p.stdin.Close()
+}
+
+// Wait waits for the process to exit. If the process has already exited returns
+// the previous error (if any).
+func (p *JobProcess) Wait() error {
+	<-p.waitBlock
+	return p.waitError
+}
+
+// Start starts the job object process
+func (p *JobProcess) Start() error {
+	return p.cmd.Start()
+}
+
+// This should only be called once.
+func (p *JobProcess) waitBackground(ctx context.Context) {
+	log.G(ctx).WithField("pid", p.Pid()).Debug("waitBackground for JobProcess")
+
+	var err error
+	if p.cmd.Process == nil {
+		err = errors.New("process has not been started")
+	}
+
+	// Wait for process to get signaled/exit/terminate/.
+	err = p.cmd.Wait()
+
+	// Wait closes the stdio pipes so theres no need to later on.
+	p.stdioLock.Lock()
+	p.stdin = nil
+	p.stdout = nil
+	p.stderr = nil
+	p.stdioLock.Unlock()
+
+	p.closedWaitOnce.Do(func() {
+		p.waitError = err
+		close(p.waitBlock)
+	})
+}
+
+// ExitCode returns the exit code of the process. The process must have
+// already terminated.
+func (p *JobProcess) ExitCode() (int, error) {
+	if !p.exited() {
+		return -1, errors.New("process has not exited")
+	}
+	return p.cmd.ProcessState.ExitCode(), nil
+}
+
+// Pid returns the processes PID
+func (p *JobProcess) Pid() int {
+	if process := p.cmd.Process; process != nil {
+		return process.Pid
+	}
+	return 0
+}
+
+// Close cleans up any state associated with the process but does not kill it.
+func (p *JobProcess) Close() error {
+	p.stdioLock.Lock()
+	if p.stdin != nil {
+		p.stdin.Close()
+		p.stdin = nil
+	}
+	if p.stdout != nil {
+		p.stdout.Close()
+		p.stdout = nil
+	}
+	if p.stderr != nil {
+		p.stderr.Close()
+		p.stderr = nil
+	}
+	p.stdioLock.Unlock()
+
+	p.closedWaitOnce.Do(func() {
+		p.waitError = nil
+		close(p.waitBlock)
+	})
+	return nil
+}
+
+// Kill kills the running process. Go calls TerminateProcess under the hood.
+func (p *JobProcess) Kill(ctx context.Context) (bool, error) {
+	log.G(ctx).WithField("pid", p.Pid()).Debug("killing job process")
+
+	if p.exited() {
+		return false, errors.New("kill not sent. process already exited")
+	}
```

Comment:
Why have this outside the lock? Do you know if setting the process state to exited is done in a lock internally?